### PR TITLE
wmb apis (notification wip)

### PR DESCRIPTION
**Pull request by 00abhishek0000**
## Type of Change

- [ ] Bugfix
- [ ] New feature
- [ ] Enhancement
- [ ] Refactoring
- [ ] Dependency updates

### Additional Changes

- [ ] This PR modifies the database schema (database migration added)
- [ ] This PR modifies dhall configs/environment variables
- [ ] This PR contains API breaking changes

## Checklist

- [ ] I formatted the code and addressed linter errors `./dev/format-all-files.sh`
- [ ] I reviewed submitted code
- [ ] I added unit tests for my changes where possible
- [ ] I added integration tests for my changes where possible
- [ ] I tested the changes using integration tests
- [ ] I added a [CHANGELOG](/CHANGELOG.md) entry if applicable
- [ ] No leak detected in leakcanary

## Release Notes

- **New Features**
  - Added trip transaction management capabilities.
  - Introduced trip planning and tracking functionality.
  - Enhanced route and vehicle mapping with encryption support.
  - Added new API endpoints for deleting trips and managing trip transactions.
  - Implemented a new approval request system for trip management.

- **Improvements**
  - Added support for trip status tracking (assigned, in progress, paused, completed, cancelled).
  - Implemented fleet configuration management.
  - Improved driver profile and route management.
  - Enhanced query capabilities for vehicle route mappings and trip transactions.

- **Bug Fixes**
  - Updated vehicle route mapping to support encrypted vehicle numbers.
  - Enhanced database query performance with new indexes.

- **Security**
  - Added encryption for vehicle numbers.
  - Implemented data protection mechanisms for sensitive information.

### CODE REVIEW

File: Backend/app/provider-platform/dynamic-offer-driver-app/Main/src/Domain/Action/UI/WMB.hs [hunk 1]
```diff
@@ -31,26 +101,171 @@ getUiWmbAvailableRoutes ::
...
+  let dataSend =
+        TN.WMBTripAssignedData
+          { tripTransactionId = tripTransactionId,
+            routeCode = obj.routeCode,
+            routeShortname = routeInfo.shortName,
+            vehicleNumber = obj.vehicleNumber,
+            vehicleServiceTierType = mapping.vehicleServiceTierType,
+            roundRouteCode = routeInfo.roundRouteCode
+          }
+  TN.notifyWmbOnRideAssigned driverId city "TRIP_ASSIGNED" "Ride Assigned" "You have been assigned a ride" dataSend
```
Comment (piyushKumar-1) [inline Backend/app/provider-platform/dynamic-offer-driver-app/Main/src/Domain/Action/UI/WMB.hs:182]: use enum here for status

**Comment by max-keviv** [comment]
@00abhishek0000 we will need status in trip list api

File: Backend/app/provider-platform/dynamic-offer-driver-app/Main/spec/Storage/FRFSStaticData.yaml [hunk 2]
```diff
@@ -173,4 +173,11 @@ RouteTripStopMapping:
     routeCode: PrimaryKey
     stopCode: PrimaryKey
     tripSequenceNum: PrimaryKey
-    scheduledDay: PrimaryKey
\ No newline at end of file
+    scheduledDay: PrimaryKey
+  queries:
+    findByLocation:
+      kvFunction: findOneWithKV
+      where: stopPoint
```
[the remark this answers is not in the document]
Comment (khuzema786) [reply Backend/app/provider-platform/dynamic-offer-driver-app/Main/spec/Storage/FRFSStaticData.yaml:180]: @00abhishek0000 Why is this Query required on stopPoint ?

File: Backend/app/provider-platform/dynamic-offer-driver-app/Main/spec/Storage/Merchant.yaml [hunk 3]
```diff
@@ -111,7 +111,7 @@ Merchant:
     registryUrl: "'http://localhost:8020'"
     uniqueKeyId: "'FIXME'"
     onlinePayment: "false"
-    gatewayAndRegistryPriorityList: "'{\"NY\"}'"
+    gatewayAndRegistryPriorityList: '''{"NY"}'''
```
Comment (khuzema786) [inline Backend/app/provider-platform/dynamic-offer-driver-app/Main/spec/Storage/Merchant.yaml:114]: Why this change ? can be reverted.

File: Backend/app/provider-platform/dynamic-offer-driver-app/Main/spec/Storage/Merchant.yaml [hunk 4]
```diff
@@ -593,8 +593,8 @@ MerchantServiceUsageConfig:
     getTripRoutes: "'Google'"
     issueTicketService: "'Kapture'"
     getExophone: "'Exotel'"
-    snapToRoadProvidersList: "'{\"OSRM\", \"Google\"}'"
-    verificationProvidersPriorityList: "'{\"Idfy\"}'"
+    snapToRoadProvidersList: '''{"OSRM", "Google"}'''
```
Comment (khuzema786) [inline Backend/app/provider-platform/dynamic-offer-driver-app/Main/spec/Storage/Merchant.yaml:596]: Revert these changes to already existing fields.

File: Backend/app/provider-platform/dynamic-offer-driver-app/Main/spec/Storage/Merchant.yaml [hunk 5]
```diff
@@ -1096,6 +1097,7 @@ TransporterConfig:
     issueBreachConfig: Kernel.Prelude.toJSON <$> issueBreachConfig|E
     recentScheduledBookingsSafeLimit: Just recentScheduledBookingsSafeLimit|E
     minDistanceForStopFcm: Just minDistanceForStopFcm|E
+    endRideDistanceThreshold: Just endRideDistanceThreshold|E
```
Comment (khuzema786) [inline Backend/app/provider-platform/dynamic-offer-driver-app/Main/spec/Storage/Merchant.yaml:1100]: Move this to FleetConfig

File: Backend/app/provider-platform/dynamic-offer-driver-app/Main/src/Domain/Action/UI/DriverProfileQuestions.hs [hunk 6]
```diff
@@ -104,7 +104,7 @@ postDriverProfileQues (mbPersonId, merchantId, merchantOpCityId) req@API.Types.U
     genAboutMeWithAI person driverStats now req' = do
       orgLLMChatCompletionConfig <- QOMC.findByMerchantOpCityId merchantOpCityId Nothing >>= fromMaybeM (MerchantServiceUsageConfigNotFound merchantOpCityId.getId)
       prompt <-
-        SCL.findByMerchantOpCityIdAndServiceNameAndUseCaseAndPromptKey merchantOpCityId (DOSC.LLMChatCompletionService $ (.llmChatCompletion) orgLLMChatCompletionConfig) DTL.DriverProfileGen DTL.DriverProfileGen_1 >>= fromMaybeM (LlmPromptNotFound merchantOpCityId.getId (show (DOSC.LLMChatCompletionService $ (.llmChatCompletion) orgLLMChatCompletionConfig)) (show DTL.DriverProfileGen) (show DTL.DriverProfileGen_1))
+        SCL.findByMerchantOpCityIdAndServiceNameAndUseCaseAndPromptKey merchantOpCityId (DOSC.LLMChatCompletionService $ (.llmChatCompletion) orgLLMChatCompletionConfig) DTL.DriverProfileGen DTL.AzureOpenAI_DriverProfileGen_1 >>= fromMaybeM (LlmPromptNotFound merchantOpCityId.getId (show (DOSC.LLMChatCompletionService $ (.llmChatCompletion) orgLLMChatCompletionConfig)) (show DTL.DriverProfileGen) (show DTL.AzureOpenAI_DriverProfileGen_1))
```
Comment (khuzema786) [inline Backend/app/provider-platform/dynamic-offer-driver-app/Main/src/Domain/Action/UI/DriverProfileQuestions.hs:107]: Why this change ? Can be reverted.

File: Frontend/ui-common/src/RemoteConfig/Utils.purs [hunk 7]
```diff
@@ -278,14 +278,14 @@ defaultFreeTrialLottie = {
 
 defaultIntroductoryLottie :: LanguageKeyValue
 defaultIntroductoryLottie = {
-  english: "lottie/ny_sub_intro_english_jan1.json",
-  hindi: "lottie/ny_sub_intro_hindi_jan1.json",
-  kannada: "lottie/ny_sub_intro_kannada_jan1.json",
-  tamil: "lottie/ny_sub_intro_tamil_jan1.json",
+  english: "lottie/ny_sub_intro_english_apr1.json",
```
Comment (khuzema786) [inline Frontend/ui-common/src/RemoteConfig/Utils.purs:281]: Why this changes ?

File: Backend/app/dashboard/CommonAPIs/spec/ProviderPlatform/Fleet/API/Driver.yaml [hunk 8]
```diff
@@ -467,7 +468,44 @@ apis:
...
+  - PUT:
+      endpoint: /dashboard/fleet/wmb/trip/{tripTransactionId}/delete
+      auth: ApiAuthV2
+      params:
+        tripTransactionId: Id TripTransaction
+      response:
+        type: APISuccess
+      helperApi:
+        - PUT:
+            endpoint: /dashboard/fleet/wmb/trip/{tripTransactionId}/{fleetOwnerId}/delete
```
Comment (piyushKumar-1) [inline Backend/app/dashboard/CommonAPIs/spec/ProviderPlatform/Fleet/API/Driver.yaml:480]: `/dashboard/fleet/wmb/trip/{tripTransactionId}/{fleetOwnerId}/delete` -> `/dashboard/fleet/{fleetOwnerId}/wmb/trip/{tripTransactionId}/delete`

File: Backend/app/dashboard/CommonAPIs/src/Dashboard/Common.hs [hunk 9]
```diff
@@ -11,6 +11,9 @@
 
  the GNU Affero General Public License along with this program. If not, see <https://www.gnu.org/licenses/>.
 -}
+{-# LANGUAGE DerivingStrategies #-}
+{-# LANGUAGE GeneralizedNewtypeDeriving #-}
+{-# LANGUAGE TemplateHaskell #-}
```
Comment (piyushKumar-1) [inline Backend/app/dashboard/CommonAPIs/src/Dashboard/Common.hs:16]: are these required ?

File: Backend/app/provider-platform/dynamic-offer-driver-app/Main/spec/Storage/WhereIsMyBus.yaml [hunk 10]
```diff
@@ -70,25 +82,62 @@ TripTransaction:
...
+      lon: Double
+      reason: Maybe Text
+      tripTransactionId: Id TripTransaction
+      derive: "Eq,Ord,Read"
 
   fields:
-    id: Id DriverRequest
-    requestType: Text
-    description: Maybe Text
-    tripTransactionId : Id TripTransaction
```
Comment (piyushKumar-1) [inline Backend/app/provider-platform/dynamic-offer-driver-app/Main/spec/Storage/WhereIsMyBus.yaml:84]: without tripTransactionId, how will we link which request is approved for which driver and which trip ?

File: Backend/app/provider-platform/dynamic-offer-driver-app/Main/src/Domain/Action/UI/WMB.hs [hunk 11]
```diff
@@ -6,51 +6,299 @@ module Domain.Action.UI.WMB
...
       Kernel.Types.Id.Id Domain.Types.MerchantOperatingCity.MerchantOperatingCity
     ) ->
     API.Types.UI.WMB.TripLinkReq ->
-    Environment.Flow API.Types.UI.WMB.TripLinkResp
+    Environment.Flow API.Types.UI.WMB.TripTransactionDetails
   )
-postUiWmbTripLink = do error "Logic yet to be decided"
+postUiWmbTripLink (person, mer, city) obj = do
+  driverId <- fromMaybeM (InternalError "Driver ID not found") person
+  fleetDriverAssociation <- FDA.findByDriver driverId >>= fromMaybeM (InternalError "Fleet Association not found")
```
Comment (piyushKumar-1) [inline Backend/app/provider-platform/dynamic-offer-driver-app/Main/src/Domain/Action/UI/WMB.hs:131]: check the comment above, we don't want to fail this API is a driver is not linked to any fleet, just link the driver to the fleet to which the vehicleHash in `TripLinkReq` belongs

File: Backend/app/provider-platform/dynamic-offer-driver-app/Main/src/Domain/Action/UI/WMB.hs [hunk 12]
```diff
@@ -6,51 +6,299 @@ module Domain.Action.UI.WMB
...
   ( ( Kernel.Prelude.Maybe (Kernel.Types.Id.Id Domain.Types.Person.Person),
       Kernel.Types.Id.Id Domain.Types.Merchant.Merchant,
       Kernel.Types.Id.Id Domain.Types.MerchantOperatingCity.MerchantOperatingCity
     ) ->
-    Environment.Flow API.Types.UI.WMB.TripLinkResp
+    Environment.Flow API.Types.UI.WMB.ActiveTripTransaction
   )
-getUiWmbTripActive = do error "Logic yet to be decided"
+getUiWmbTripActive (person, _, _) = do
+  checkFleetDriverAssociation person >>= \isAssociated -> unless isAssociated (throwError $ InternalError "Fleet-driver association not found")
```
Comment (piyushKumar-1) [inline Backend/app/provider-platform/dynamic-offer-driver-app/Main/src/Domain/Action/UI/WMB.hs:278]: ```unlessM `checkFleetDriverAssociation personId` . throwError $ InternalError "Fleet-driver association not found"```

File: Backend/app/provider-platform/dynamic-offer-driver-app/Main/src/Domain/Action/UI/WMB.hs [hunk 13]
```diff
@@ -61,7 +309,19 @@ postUiWmbTripStart ::
     API.Types.UI.WMB.TripStartReq ->
     Environment.Flow Kernel.Types.APISuccess.APISuccess
   )
-postUiWmbTripStart = do error "Logic yet to be decided"
+postUiWmbTripStart (person, _, city) tripTransactionId obj = do
```
Comment (piyushKumar-1) [inline Backend/app/provider-platform/dynamic-offer-driver-app/Main/src/Domain/Action/UI/WMB.hs:312]: name it personId instead of person also `obj` -> `req` ([good read on proper naming](https://dev.to/gervaisamoah/a-guide-to-clean-code-the-power-of-good-names-3f6i?utm_source=chatgpt.com))

File: Backend/app/provider-platform/dynamic-offer-driver-app/Main/src/Domain/Action/UI/WMB.hs [hunk 14]
```diff
@@ -6,51 +6,299 @@ module Domain.Action.UI.WMB
...
+--   let toRadians deg = deg * pi / 180
+--       r = 6371.0
+--       dlat = toRadians (lat2 - lat1)
+--       dlon = toRadians (lon2 - lon1)
+--       a = sin (dlat / 2) * sin (dlat / 2) + cos (toRadians lat1) * cos (toRadians lat2) * sin (dlon / 2) * sin (dlon / 2)
+--       c = 2 * Kernel.Prelude.atan2 (sqrt a) (sqrt (1 - a))
+--    in r * c
+
+findClosestStop :: LatLong -> [(Text, LatLong)] -> Maybe Text
+findClosestStop loc stops =
```
Comment (piyushKumar-1) [inline Backend/app/provider-platform/dynamic-offer-driver-app/Main/src/Domain/Action/UI/WMB.hs:298]: what if its equidistant from two stops ? we should consider stop with higher sequence number in that case.

File: Backend/app/provider-platform/dynamic-offer-driver-app/Main/src/Domain/Action/UI/WMB.hs [hunk 15]
```diff
@@ -72,4 +332,67 @@ postUiWmbTripEnd ::
...
+  driverId <- fromMaybeM (InternalError "Driver ID not found") person
+  fleetDriverAssociation <- FDA.findByDriver driverId >>= fromMaybeM (InternalError "Fleet Association not found")
+  transaction <- QTT.findByTransactionId (Kernel.Types.Id.Id tripTransactionId) >>= fromMaybeM (InternalError "no trip transaction found")
+  routeInfo <- QRM.findByRouteCode transaction.routeCode >>= fromMaybeM (InternalError "Route not found")
+  transporterConfig <- QTC.findByMerchantOpCityId merch_city_id >>= fromMaybeM (InternalError "Config not found")
+  fleetConfigInfo <- QFC.findByPrimaryKey (Kernel.Types.Id.Id fleetDriverAssociation.fleetOwnerId) >>= fromMaybeM (InternalError "Fleet Config Info not found")
+  now <- getCurrentTime
+
+  let distanceLeft = KU.distanceBetweenInMeters obj.location routeInfo.endPoint
+      updateAndRespond response = do
```
Comment (piyushKumar-1) [inline Backend/app/provider-platform/dynamic-offer-driver-app/Main/src/Domain/Action/UI/WMB.hs:345]: its better to keep these kind of helper functions in where clause for better readability

File: Backend/app/provider-platform/dynamic-offer-driver-app/Main/src/Domain/Action/UI/WMB.hs [hunk 16]
```diff
@@ -72,4 +332,67 @@ postUiWmbTripEnd ::
...
+                  merchantId = Nothing,
+                  merchantOperatingCityId = Just merch_city_id
+                }
+        QDR.create driver_req
+        pure $ TripEndResp {requestId = Just requestId.getId, result = WAITING_FOR_ADMIN_APPROVAL}
+
+  if fleetConfigInfo.allowEndingMidRoute || distanceLeft <= transporterConfig.endRideDistanceThreshold
+    then updateAndRespond $ TripEndResp {requestId = Nothing, result = SUCCESS}
+    else
+      if fleetConfigInfo.rideEndApproval
```
Comment (piyushKumar-1) [inline Backend/app/provider-platform/dynamic-offer-driver-app/Main/src/Domain/Action/UI/WMB.hs:379]: `rideEndApproval` not clear if this means its required or notRequired, can be mistaken later.

**Review by khuzema786 — APPROVED** [review]

**Review by khuzema786 — APPROVED** [review]
